File: androidbase-sdk/src/main/java/com/leovp/androidbase/ui/FloatViewManager.java
package com.leovp.androidbase.ui;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.graphics.PixelFormat;
import android.os.Build;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

import androidx.annotation.LayoutRes;

/**
 * Author: Michael Leo
 * Date: 20-3-3 下午3:40
 *
 * @see <a href="https://stackoverflow.com/a/53092436">Float View</a>
 */
public class FloatViewManager {
    private static boolean mIsFloatViewShowing;

    private Activity mActivity;
    private WindowManager mWindowManager;
    private View mFloatView;
    private WindowManager.LayoutParams mFloatViewLayoutParams;
    private boolean mFloatViewTouchConsumedByMove = false;
    private int mFloatViewLastX;
    private int mFloatViewLastY;
    private int mFloatViewFirstX;
    private int mFloatViewFirstY;

    private View.OnClickListener clickListener;

    @SuppressLint("InflateParams")
    public FloatViewManager(Activity activity, @LayoutRes int layoutId, View.OnClickListener clickListener) {
        mActivity = activity;
        this.clickListener = clickListener;
        LayoutInflater inflater = LayoutInflater.from(activity);
        mFloatView = inflater.inflate(layoutId, null);
        mFloatView.setOnTouchListener(mFloatViewOnTouchListener);

        mFloatViewLayoutParams = new WindowManager.LayoutParams();
        mFloatViewLayoutParams.format = PixelFormat.TRANSLUCENT;
        mFloatViewLayoutParams.flags = WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE;

        mFloatViewLayoutParams.type = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
                ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                : WindowManager.LayoutParams.TYPE_TOAST;

        mFloatViewLayoutParams.gravity = Gravity.CENTER;
        mFloatViewLayoutParams.width = WindowManager.LayoutParams.WRAP_CONTENT;
        mFloatViewLayoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
    }

    public void dismissFloatView() {
        if (mIsFloatViewShowing) {
            mIsFloatViewShowing = false;
            mActivity.runOnUiThread(() -> {
                if (mWindowManager != null) {
                    mWindowManager.removeViewImmediate(mFloatView);
                }
            });
        }
    }

    public void showFloatView() {
        if (!mIsFloatViewShowing) {
            mIsFloatViewShowing = true;
            mActivity.runOnUiThread(() -> {
                if (!mActivity.isFinishing()) {
                    mWindowManager = (WindowManager) mActivity.getSystemService(Context.WINDOW_SERVICE);
                    if (mWindowManager != null) {
                        mWindowManager.addView(mFloatView, mFloatViewLayoutParams);
                    }
                }
            });
        }
    }

    @SuppressWarnings("FieldCanBeLocal")
    private View.OnTouchListener mFloatViewOnTouchListener = new View.OnTouchListener() {

        @SuppressLint("ClickableViewAccessibility")
        @TargetApi(Build.VERSION_CODES.FROYO)
        @Override
        public boolean onTouch(View v, MotionEvent event) {
            WindowManager.LayoutParams prm = mFloatViewLayoutParams;
            int totalDeltaX = mFloatViewLastX - mFloatViewFirstX;
            int totalDeltaY = mFloatViewLastY - mFloatViewFirstY;

            switch (event.getActionMasked()) {
                case MotionEvent.ACTION_DOWN:
                    mFloatViewLastX = (int) event.getRawX();
                    mFloatViewLastY = (int) event.getRawY();
                    mFloatViewFirstX = mFloatViewLastX;
                    mFloatViewFirstY = mFloatViewLastY;
                    break;
                case MotionEvent.ACTION_UP:
                    if (Math.abs(totalDeltaX) < 5 || Math.abs(totalDeltaY) < 5) {
                        if (event.getPointerCount() == 1) {
                            clickListener.onClick(null);
                        }
                    }
                    break;
                case MotionEvent.ACTION_MOVE:
                    int deltaX = (int) event.getRawX() - mFloatViewLastX;
                    int deltaY = (int) event.getRawY() - mFloatViewLastY;
                    mFloatViewLastX = (int) event.getRawX();
                    mFloatViewLastY = (int) event.getRawY();
                    if (Math.abs(totalDeltaX) >= 5 || Math.abs(totalDeltaY) >= 5) {
                        if (event.getPointerCount() == 1) {
                            prm.x += deltaX;
                            prm.y += deltaY;
                            mFloatViewTouchConsumedByMove = true;
                            if (mWindowManager != null) {
                                mWindowManager.updateViewLayout(mFloatView, prm);
                            }
                        } else {
                            mFloatViewTouchConsumedByMove = false;
                        }
                    } else {
                        mFloatViewTouchConsumedByMove = false;
                    }
                    break;
                default:
                    break;
            }
            return mFloatViewTouchConsumedByMove;
        }
    };

}

File: androidbase-sdk/src/main/java/com/leovp/androidbase/ui/FloatViewManager.kt
package com.leovp.androidbase.ui

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Context.WINDOW_SERVICE
import android.graphics.PixelFormat
import android.os.Build
import android.view.*
import androidx.annotation.LayoutRes
import androidx.core.view.children
import com.leovp.androidbase.exts.android.*
import com.leovp.androidbase.exts.kotlin.fail
import com.leovp.androidbase.utils.log.LogContext
import kotlin.math.abs


/**
 * Author: Michael Leo
 * Date: 20-3-3 下午3:40
 *
 * Need permission: `<uses-permission android:name="android.permission.SYSTEM_ALERT_WINDOW" />`
 *
 * @see [Float View](https://stackoverflow.com/a/53092436)
 * @see [Float View Github](https://github.com/aminography/FloatingWindowApp)
 */
class FloatViewManager(
    private val context: Context,
    @LayoutRes private var layoutId: Int,
    private val posX: Int = 0,
    private val posY: Int = 0
) {
    var touchEventListener: TouchEventListener? = null

    private var _enableDrag = true

    private var _fullScreenFloatView = false
    private var _canDragOverStatusBar = false

    /**
     * If [_canDragOverStatusBar] is false, this values is the height of status bar.
     */
    private var drawHeightOffset = 0

    var enableDrag: Boolean
        get() = _enableDrag
        set(enableDrag) {
            _enableDrag = enableDrag
            updateLayout()
        }

    /**
     * Whether the window itself in full screen mode.
     * In full screen mode, the [enableDrag] will be ignored and the click event for this float view itself will be consumed.
     */
    var enableFullScreenFloatView: Boolean
        get() = _fullScreenFloatView
        set(enableFloatViewInFullScreen) {
            _fullScreenFloatView = enableFloatViewInFullScreen
            updateLayout()
        }

    /**
     * If the float view is in full screen mode, this property will be ignored.
     */
    var canDragOverStatusBar: Boolean
        get() = _canDragOverStatusBar
        set(canDragOverStatusBar) {
            _canDragOverStatusBar = canDragOverStatusBar
            updateLayout()
        }

    var x: Int
        get() = layoutParams.x
        set(x) {
            if (context !is Activity) fail("The context should be Activity context.")
            LogContext.log.e("floatView dimension x=${floatView.width}x${floatView.height}")
            var finalX = x
            if (x < 0) finalX = 0
            if ((x + layoutParams.width) > context.screenRealWidth) finalX = context.screenRealWidth - layoutParams.width
            layoutParams.x = finalX
            windowManager.updateViewLayout(floatView, layoutParams)

//            if (GlobalConstants.DEBUG) LogContext.log.e( "floatView dimension55=${floatView.width}x${floatView.height}")
//            if (GlobalConstants.DEBUG) LogContext.log.e( "floatView dimension66=${floatView.measuredWidth}x${floatView.measuredHeight}")
//            Handler(context.mainLooper).post {
//                if (GlobalConstants.DEBUG) LogContext.log.e( "floatView dimension77=${floatView.width}x${floatView.height}")
//                if (GlobalConstants.DEBUG) LogContext.log.e( "floatView dimension88=${floatView.measuredWidth}x${floatView.measuredHeight}")
//            }
        }

    var y: Int
        get() = layoutParams.y
        set(y) {
            LogContext.log.e("floatView dimension y=${floatView.width}x${floatView.height}")
            var finalY = y
            if (y < drawHeightOffset) finalY = drawHeightOffset
            if ((y + layoutParams.height) > context.screenAvailableHeight) finalY = context.screenAvailableHeight - layoutParams.height
            layoutParams.y = finalY
            windowManager.updateViewLayout(floatView, layoutParams)
        }

    /**
     * Note that, [dismiss] method will be executed when you call this method.
     */
    fun replaceLayout(@LayoutRes layoutId: Int) {
        dismiss()
        this.layoutId = layoutId
        floatView = LayoutInflater.from(context).inflate(layoutId, null)
        init()
    }

    var floatView: View = LayoutInflater.from(context).inflate(layoutId, null)
        private set

    @SuppressLint("ClickableViewAccessibility")
    private val onTouchListener = View.OnTouchListener { view, event ->
        if (!_enableDrag || _fullScreenFloatView) return@OnTouchListener true
        val totalDeltaX = lastX - firstX
        val totalDeltaY = lastY - firstY

        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                lastX = event.rawX.toInt()
                lastY = event.rawY.toInt()
                firstX = lastX
                firstY = lastY
                isClickGesture = true
//                if (GlobalConstants.DEBUG) LogContext.log.e("ACTION_DOWN isPressed=${view.isPressed} hasFocus=${view.hasFocus()} isActivated=${view.isActivated} $view")
                touchConsumedByMove = touchEventListener?.touchDown(view, lastX, lastY) ?: false
            }
            MotionEvent.ACTION_UP -> {
//                view.performClick()
//                if (GlobalConstants.DEBUG) LogContext.log.e("ACTION_UP isPressed=${view.isPressed} hasFocus=${view.hasFocus()} isActivated=${view.isActivated} isClickGesture=$isClickGesture $view")
                touchConsumedByMove = touchEventListener?.touchUp(view, lastX, lastY, isClickGesture) ?: !isClickGesture
            }
            MotionEvent.ACTION_MOVE -> {
                val deltaX = event.rawX.toInt() - lastX
                val deltaY = event.rawY.toInt() - lastY
                lastX = event.rawX.toInt()
                lastY = event.rawY.toInt()
                if (abs(totalDeltaX) >= TOUCH_TOLERANCE_IN_PX || abs(totalDeltaY) >= TOUCH_TOLERANCE_IN_PX) {
                    isClickGesture = false
                    view.isPressed = false
                    if (event.pointerCount == 1) {
                        layoutParams.x += deltaX
                        layoutParams.y += deltaY
                        if (layoutParams.y <= drawHeightOffset) layoutParams.y = drawHeightOffset
                        touchConsumedByMove = true
                        windowManager.updateViewLayout(floatView, layoutParams)
                    } else {
                        touchConsumedByMove = false
                    }
                } else {
                    isClickGesture = true
                    touchConsumedByMove = false
                }
                touchConsumedByMove = touchEventListener?.touchMove(view, lastX, lastY, isClickGesture) ?: touchConsumedByMove
            }
            else -> Unit
        }
        touchConsumedByMove
    }

    init {
        init()
    }

    private fun init() {
        updateLayout()
        // Ensure to set the touch listener to the root view.
        floatView.setOnTouchListener(onTouchListener)
    }

    private fun addTouchListenerToView(view: View, touchListener: View.OnTouchListener) {
        if (view is ViewGroup) {
//            runCatching {
//                if (GlobalConstants.DEBUG) LogContext.log.e("Found ViewGroup: ${app.resources.getResourceEntryName(view.id)}:${view::class.simpleName}")
//            }
            if (view.hasOnClickListeners()) {
//                runCatching {
//                    if (GlobalConstants.DEBUG) LogContext.log.e("setOnTouchListener for ViewGroup: ${app.resources.getResourceEntryName(view.id)}:${view::class.simpleName}")
//                }
                view.setOnTouchListener(touchListener)
            }
            view.children.forEach { child -> addTouchListenerToView(child, touchListener) }
        } else {
//            runCatching {
//                if (GlobalConstants.DEBUG) LogContext.log.e("Found View     : ${app.resources.getResourceEntryName(view.id)}:${view::class.simpleName}")
//            }
            if (view.hasOnClickListeners()) {
//                runCatching {
//                    if (GlobalConstants.DEBUG) LogContext.log.e("setOnTouchListener for View     : ${app.resources.getResourceEntryName(view.id)}:${view::class.simpleName}")
//                }
                view.setOnTouchListener(touchListener)
            }
        }
    }

    private val windowManager: WindowManager = (context.getSystemService(WINDOW_SERVICE) as WindowManager)

    private lateinit var layoutParams: WindowManager.LayoutParams

    private var lastX: Int = 0
    private var lastY: Int = 0
    private var firstX: Int = 0
    private var firstY: Int = 0

    var isShowing = false
        private set
    private var touchConsumedByMove = false

    private var isClickGesture = true

    private fun updateLayout() {
        drawHeightOffset = if (_canDragOverStatusBar) 0 else context.statusBarHeight
        layoutParams = WindowManager.LayoutParams().apply {
            format = PixelFormat.TRANSLUCENT
            flags = if (_enableDrag && !_fullScreenFloatView) {
                WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE // or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
            } else {
                // FLAG_NOT_TOUCHABLE will bubble the event to the bottom layer.
                // However the float layer itself can not be touched anymore.
                WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN or WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE // or WindowManager.LayoutParams.FLAG_LAYOUT_INSET_DECOR
            }
            @Suppress("DEPRECATION")
            type = when {
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                else -> WindowManager.LayoutParams.TYPE_TOAST
            }

            gravity = Gravity.TOP or Gravity.START // Default value: Gravity.CENTER
            x = posX
            y = posY
            width = if (_fullScreenFloatView) WindowManager.LayoutParams.MATCH_PARENT else WindowManager.LayoutParams.WRAP_CONTENT
            height = if (_fullScreenFloatView) WindowManager.LayoutParams.MATCH_PARENT else WindowManager.LayoutParams.WRAP_CONTENT
        }
    }

    fun show() {
        if (context.canDrawOverlays) {
            runCatching {
                dismiss()
                addTouchListenerToView(floatView, onTouchListener)
                isShowing = true
                windowManager.addView(floatView, layoutParams)
            }.onFailure { it.printStackTrace() }
        }
    }

    @Suppress("unused")
    fun dismiss() {
        if (isShowing) {
            windowManager.removeView(floatView)
            isShowing = false
        }
    }

    interface TouchEventListener {
        fun touchDown(view: View, x: Int, y: Int): Boolean = false
        fun touchMove(view: View, x: Int, y: Int, isClickGesture: Boolean): Boolean = true

        /**
         * Generally, if [isClickGesture] is `false` that means before touch up being triggered, user is just moving the view.
         * At this time, we should consume this touch event so the click listener that use set should NOT be triggered.
         *
         * In contrast, if [isClickGesture] is `true` that means user triggers the click event,
         * so this touch event should not be consumed.
         */
        fun touchUp(view: View, x: Int, y: Int, isClickGesture: Boolean): Boolean = !isClickGesture
    }

    companion object {
        private const val TOUCH_TOLERANCE_IN_PX = 8
    }
}